uuid.go

```diff
@@ -64,6 +67,33 @@ const (
 	DomainOrg
 )
 
+// Timestamp is the count of 100-nanosecond intervals since 00:00:00.00,
+// 15 October 1582 within a V1 UUID. This type has no meaning for V2-V5
+// UUIDs since they don't have an embedded timestamp.
+type Timestamp uint64
+
+const _100nsPerSecond = 10000000
+
+// Time returns the UTC time.Time representation of a Timestamp
+func (t Timestamp) Time() (time.Time, error) {
+	secs := uint64(t)/_100nsPerSecond
+	nsecs := 100 * (uint64(t)%_100nsPerSecond)
+	return time.Unix(int64(secs)-(epochStart/_100nsPerSecond), int64(nsecs)), nil
+}
+
+// TimestampFromV1 returns the Timestamp embedded within a V1 UUID.
+// Returns an error if the UUID is any version other than 1.
+func TimestampFromV1(u UUID) (Timestamp, error) {
+	if u.Version() != 1 {
+		err := fmt.Errorf("uuid: %s is version %d, not version 1", u, u.Version())
+		return 0, err
+	}
+	low := binary.BigEndian.Uint32(u[0:4])
+	mid := binary.BigEndian.Uint16(u[4:6])
+	hi := binary.BigEndian.Uint16(u[6:8]) & 0xfff
+	return Timestamp(low) + (Timestamp(mid) << 32) + (Timestamp(hi) << 48), nil
```

Comment:
I think using `Timestamp` type here is confusing.

How about doing math + bit shift operations with `uint64` first?

Reply:
A Timestamp is a uint64, and we need to return one. Why is it confusing?

Reply:
I'm a proponent of coding against types: I strongly believe that types carry a lot of extra context.

In this case, keeping in mind that `Timestamp` is a type that represents number of 100s of nanoseconds since `00:00:00.00 15 October 1582`, the `Timestamp(mid) << 32` operation makes very little sense: what does "shift 100s of nanoseconds 32 bits left" mean?

It's a mathematical operation and should be applied to numbers, which then (in the very end) might be interpreted as something domain-specific.

PS: it's my personal opinion, if other people are okay - I won't insist on this change too hard.

Reply:
I changed it. I don't have a strong opinion either way, but to me it seems harder to read after making the change (a ton of casting is now happening).

I suppose I could break it up into multiple lines but it still seems "wordier". I don't have a strong opinion though.

Reply:
It looks ideal to me now :-)